### PR TITLE
chore(ci): add ci workflow

### PR DESCRIPTION
**Pull request by joshua-goldstein**
Run dgo client tests using github actions.  Note that a self hosted runner is necessary because we will be using docker commands (i.e. `docker compose up`).

### CODE REVIEW

File: .github/workflows/ci-dgo-tests.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        run: |
+          #!/bin/bash
+          # clean go test cache
+          go clean -testcache
+      - name: Run dgo client tests
+        run: |
+          #!/bin/bash
+          docker compose -f t/docker-compose.yml up -d
+          echo "Waiting for cluster to be healthy..."
+          sleep 20
```
Comment (joshua-goldstein) [inline .github/workflows/ci-dgo-tests.yml:56]: Ugly, I know - but we can improve this later.

File: t/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  zero1:
+    image: dgraph/dgraph:local
+    container_name: zero1
+    working_dir: /zero1
+    ports:
+    - 5180:5180
+    - 6180:6180
+    command: dgraph zero -o 100 --raft='idx=1' --my=zero1:5180 --replicas=3
+      --logtostderr -v=2 --bindall
+volumes: {}
```
Comment (mangalaman93) [inline t/docker-compose.yml:54]: don't need this line

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
We also need to add the ci workflow for the linter. And do you like `t` name better than `contrib` for the folder?

**Review by mangalaman93 — APPROVED** [review]

File: .github/workflows/ci-dgo-tests.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - ready_for_review
+    branches:
+      - master
+  schedule:
+    - cron: "0 4 * * *" # daily at 4:00
```
Comment (skrdgraph) [inline .github/workflows/ci-dgo-tests.yml:15]: nitpick - make this `0 0 * * *` the current config would be 4am UTC

File: .github/workflows/ci-dgo-tests.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+jobs:
+  dgo-tests:
+    runs-on: [self-hosted, x64] # needed to use docker commands
```
Comment (skrdgraph) [inline .github/workflows/ci-dgo-tests.yml:18]: do you need self hosted for this?

File: .github/workflows/ci-golang-lint.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - ready_for_review
+    branches:
+      - master
+  schedule:
+    - cron: "0 4 * * *" # daily at 4:00
```
Comment (skrdgraph) [inline .github/workflows/ci-golang-lint.yml:15]: same feedback

File: t/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      -v=2 --raft "idx=3; group=1" --security whitelist=0.0.0.0/0 --acl secret-file=/data/acl_secret;access-ttl=3s
+  zero1:
+    image: dgraph/dgraph:local
+    container_name: zero1
+    working_dir: /zero1
+    ports:
+    - 5180:5180
+    - 6180:6180
+    command: dgraph zero -o 100 --raft='idx=1' --my=zero1:5180 --replicas=3
+      --logtostderr -v=2 --bindall
```
Comment (skrdgraph) [inline t/docker-compose.yml:53]: do we want to make this a HA on Zero's as well?

File: .github/workflows/ci-dgo-tests.yml (hunk 1 above)
Comment (skrdgraph) [reply .github/workflows/ci-dgo-tests.yml:56]: can you explore this https://gist.github.com/rgl/f90ff293d56dbb0a1e0f7e7e89a81f42

File: .github/workflows/ci-dgo-tests.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          path: dgraph 
+          repository: dgraph-io/dgraph
+          ref: main
+      - name: Checkout Dgo repo
+        uses: actions/checkout@v3
+        with:
+          path: dgo
+          repository: dgraph-io/dgo
+      - name: Move into Dgraph repo
+        run: cd dgraph
```
Comment (skrdgraph) [inline .github/workflows/ci-dgo-tests.yml:32]: from what I know - the CD doesn't hold true outside the sub-task. 

You would want to do this inside every sub-task

File: .github/workflows/ci-dgo-tests.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        run: cd ../dgo
+      - name: Clean Up Environment
+        run: |
+          #!/bin/bash
+          # clean go test cache
+          go clean -testcache
+      - name: Run dgo client tests
+        run: |
+          #!/bin/bash
+          docker compose -f t/docker-compose.yml up -d
```
Comment (skrdgraph) [inline .github/workflows/ci-dgo-tests.yml:54]: I have a question - why are you using a `docker-compose.yaml` ... why can't you use the `make up` command from the `contrib` folder which Matthew had written

File: t/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,53 @@
+version: "3.5"
```
Comment (skrdgraph) [inline t/docker-compose.yml:1]: I think this yaml needs to be removed, we should have a single point of reference for such things (and it should live in dgraph repo)

File: t/docker-compose.yml (hunk 6 above)
Comment (joshua-goldstein) [reply t/docker-compose.yml:53]: @mangalaman93 what is your suggestion?

File: .github/workflows/ci-dgo-tests.yml (hunk 3 above)
Comment (joshua-goldstein) [reply .github/workflows/ci-dgo-tests.yml:15]: changing to 0 0 * * *

File: .github/workflows/ci-golang-lint.yml (hunk 5 above)
Comment (joshua-goldstein) [reply .github/workflows/ci-golang-lint.yml:15]: changing to 0 0 * * *

File: .github/workflows/ci-golang-lint.yml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+          echo "GOVERSION=$GOVERSION" >> $GITHUB_ENV
+      - name: Setup Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GOVERSION }}
+      - name: golang-lint
+        env:
+          # prevent OOM
+          GOGC: 10
+        uses: golangci/golangci-lint-action@v3
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply .github/workflows/ci-golang-lint.yml:36]: Aman has a PR to bring in the yaml config, see https://github.com/dgraph-io/dgo/pull/156/files#diff-6179837f7df53a6f05c522b6b7bb566d484d5465d9894fb04910dd08bb40dcc9

It is almost identical to dgraph repo, actually I think he has enabled more linters here (e.g. gosec)

File: t/docker-compose.yml (hunk 9 above)
Comment (joshua-goldstein) [reply t/docker-compose.yml:1]: In principle I definitely agree - but the problem is that to run dgo tests, we need a cluster to be up and running.  Ideally someone should be able to checkout the dgo repo and run the tests without also checking out the dgraph repo.  You have any thoughts on how to do that?

File: .github/workflows/ci-dgo-tests.yml (hunk 8 above)
Comment (joshua-goldstein) [reply .github/workflows/ci-dgo-tests.yml:54]: If you look at the ports in this docker-compose carefully you'll see they are a little atypical from most other docker compose files (including matthew's in contrib/local-test).  These specific ports were needed to get dgo tests to run.  @mangalaman93  might have more insight about why they are using these ports

File: t/docker-compose.yml (hunk 6 above)
Comment (mangalaman93) [reply t/docker-compose.yml:53]: either way is fine. We are just doing some sanity checks, we have tests that use dgo client in dgraph repo as well.

File: .github/workflows/ci-dgo-tests.yml (hunk 8 above)
Comment (mangalaman93) [reply .github/workflows/ci-dgo-tests.yml:54]: I can change the ports, that's a small change. Could you point me to this Makefile?

**Pull request merged**